[GTK][WPE] Make Damage iterable

carlosgcampos opened the pull request:
#### 3de4a7eaf3f379b8147dd48ee98b4a24aa44684e
<pre>
[GTK][WPE] Make Damage iterable
<a href="https://bugs.webkit.org/show_bug.cgi?id=291358">https://bugs.webkit.org/show_bug.cgi?id=291358</a>

Reviewed by Nikolas Zimmermann.

Instead of exposing the internal vector we use for dirty rects, that can
contain empty rectangles after uniting, we can make Damage iterable and
always return the actual rectangles.

Canonical link: <a href="https://commits.webkit.org/296308@main">https://commits.webkit.org/296308@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a33d88f3136d3f58f79cf7a773e9781f3e01b2dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107656 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27332 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17746 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112868 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109613 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28024 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35926 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81741 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110589 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97030 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62121 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21650 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15169 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57630 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91588 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15203 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115986 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34732 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25639 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90780 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35109 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93280 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90526 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35444 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13217 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30489 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17463 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34638 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40194 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34384 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37745 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36047 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -171,75 +181,120 @@ class Damage {
+
+        const IntRect& operator*()
+        {
+            switch (m_damage->m_mode) {
```
Null-check m_damage? Or store as std::optional?

carlosgcampos replied:
Iterators can only be created with begin() or end() which always pass a valid Damage.

nikolaszimmermann replied:
Ok I usually put in an assert to document this.

nikolaszimmermann commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -171,75 +181,120 @@ class Damage {
+
+        bool operator==(const iterator& other) const
+        {
+            return m_index == other.m_index;
```
Shouldn't we also consider m_damage? Otherwise, I can create two iterators from two damage classes that could be equal.

carlosgcampos replied:
Comparing iterators from different containers is a programmer error, it's supposed to be undefined.

nikolaszimmermann replied:
Fair enough

nikolaszimmermann commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -171,75 +181,120 @@ class Damage {
+
+    private:
+        const Damage* m_damage { nullptr };
+        std::optional<BitVector::iterator> m_iter;
```
I would avoid abbreviations: m_bitVectorIterator?

carlosgcampos replied:
Well, iter is a very common abbreviation we are all familiar with.

nikolaszimmermann replied:
An iterator class having an iter member is awkward IMHO - that's why I was suggesting the name change, feel free to ignore it

nikolaszimmermann reviewed (approved):
Please check the remaining style issue to please EWS.

carlosgcampos committed:
[GTK][WPE] Make Damage iterable
https://bugs.webkit.org/show_bug.cgi?id=291358

Reviewed by Nikolas Zimmermann.

Instead of exposing the internal vector we use for dirty rects, that can
contain empty rectangles after uniting, we can make Damage iterable and
always return the actual rectangles.

Canonical link: https://commits.webkit.org/296308@main
